core/store/src/flat_state.rs

```diff
@@ -60,20 +97,79 @@ mod imp {
         }
     }
 
-    /// Possibly creates a new [`FlateState`] object backed by given storage.
-    ///
-    /// Always returns `None` if the `protocol_feature_flat_state` Cargo feature is
-    /// not enabled.  Otherwise, returns a new [`FlatState`] object backed by
-    /// specified storage if `use_flat_state` argument is true.
-    pub fn maybe_new(use_flat_state: bool, store: &Store) -> Option<FlatState> {
-        use_flat_state.then(|| FlatState { store: store.clone() })
+    /// `FlatStateFactory` provides a way to construct new flat state to pass to new tries.
+    /// It is owned by NightshadeRuntime, and thus can be owned by multiple threads, so the implementation
+    /// must be thread safe.
+    #[derive(Clone)]
+    pub struct FlatStateFactory(Arc<FlatStateFactoryInner>);
```

Comment:
This playes the same role as `ShardTries`, right? Perhaps use the same naming scheme? `ShardFlateStates` (or, if we are willing to rename, `PerShardTrie` and `PerShardFlatState`?).